model/qti/asset/handler/StimulusHandler.php

```diff
@@ -0,0 +1,189 @@
+ *
+ * @package oat\taoQtiItem\model\qti\asset\handler
+ */
+class StimulusHandler implements AssetHandler
+{
+    /**
+     * The object representation of qti item
+     *
+     * @var Item
+     */
+    protected $qtiItem;
+
+    /**
+     * The destination of the item directory
+     *
+     * @var LocalItemSource
+     */
+    protected $itemSource;
+
+    /**
+     * Applicable to all items that contain <xinclude> tag
+     *
+     * @param $relativePath
+     * @return bool
+     * @throws \common_exception_MissingParameter
+     */
+    public function isApplicable($relativePath)
+    {
+        $xincluded = array();
+        /** @var Element $xincludeElement */
+        foreach ($this->getQtiItem()->getComposingElements('oat\taoQtiItem\model\qti\Xinclude') as $xincludeElement) {
+            $xincluded[] = $xincludeElement->attr('href');
+            \common_Logger::i("Xinclude component found in resource '" .
+                $this->getQtiItem()->getIdentifier() . "' with href '" . $xincludeElement->attr('href') . "'.");
+        }
+
+        if(in_array($relativePath, $xincluded)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Store locally, in a safe directory
+     * Encoded all stimulus images to avoid file dependencies
+     *
+     * @param $absolutePath
+     * @param $relativePath
+     * @return array|mixed
+     * @throws \common_Exception
+     */
+    public function handle($absolutePath, $relativePath)
+    {
+        $safePath = $this->safePath($relativePath);
+        $this->encodeStimulusImages($absolutePath);
+
+        $info = $this->getItemSource()->add($absolutePath, basename($absolutePath), $safePath);
+        \common_Logger::i('Stimulus file \'' . $absolutePath . '\' copied.');
+
+        return $info;
+    }
+
+    /**
+     * Remove ../ to secure path
+     *
+     * @param $path
+     * @return string
+     */
+    protected function safePath($path)
+    {
+        $safePath = '';
+        if (dirname($path) !== '.') {
+            $safePath = str_replace('../', '', dirname($path)) . '/';
+        }
+        return $safePath;
+    }
+
+    /**
+     * Walk into stimulus file to transform images from path to base64encoded data:image
+     *
+     * @param $absolutePath
+     */
+    protected function encodeStimulusImages($absolutePath)
+    {
+        $dom = new \DOMDocument('1.0', 'UTF-8');
+        $dom->loadXML(file_get_contents($absolutePath));
+        $images = $dom->getElementsByTagName('img');
+
+        for($i=0 ; $i<$images->length ; $i++) {
+            $imageFile = dirname($absolutePath) . DIRECTORY_SEPARATOR . ltrim($images->item($i)->getAttribute('src'), DIRECTORY_SEPARATOR);
+            if (file_exists($imageFile)) {
+                $encodedSrc = 'data:image/png;base64,' . base64_encode(file_get_contents($imageFile));
+                $images->item($i)->setAttribute('src', $encodedSrc);
+            }
+        }
+        if (isset($encodedSrc)) {
```

Comment:
and here and/or `is_writable`

Reply:
Check done at the beginning of the method to avoid process if it is not writable

Reply:
indeed :)